### PR TITLE
Differentiate quality from mole fraction in Python

### PR DESCRIPTION
**Pull request by ischoegl**
Please fill in the issue number this pull request is fixing:

Fixes: N/A; but follows discussion in review of #687:

> > [...] it may make sense to use Q for quality at some point in the future. [...]

> This would also be consistent with other naming conventions that have been adopted in the Python module, e.g. D for density instead of R for Rho. [...]

Changes proposed in this pull request:
- Use `Q` instead of `X` for quality (vapor fraction) 
- Deprecate all instances using `X` in `PureFluid`
- ~In `SolutionArray.collect`, the species name is appended in labels, i.e. `Q_H2O`~

I.e.
```python
In [1]: import cantera as ct

In [2]: w = ct.Water()

In [3]: w.TX = 400, .5
DeprecationWarning: To be removed after Cantera 2.5. Attribute renamed to 'TQ'

In [4]: w.TQ
Out[4]: (400.0, 0.5)
```

### CODE REVIEW

File: interfaces/cython/cantera/thermo.pyx [hunk 1]
```diff
@@ -1440,40 +1453,93 @@ cdef class PureFluid(ThermoPhase):
     """
 
     _full_states = {frozenset(k): k
-                    for k in ('TD', 'TP', 'UV', 'DP', 'HP', 'SP', 'SV', 'TX',
-                              'PX', 'ST', 'TV', 'PV', 'UP', 'VH', 'TH', 'SH')}
+                    for k in ('TD', 'TP', 'UV', 'DP', 'HP', 'SP', 'SV', 'TQ',
```
Comment (ischoegl) [inline interfaces/cython/cantera/thermo.pyx]: I am wondering whether `TP`, `TQ` and `PQ` should be removed from here, as they are not necessarily full states.

File: interfaces/cython/cantera/composite.py [hunk 2]
```diff
@@ -631,14 +633,15 @@ def restore_data(self, data, labels):
...
-                           if fs not in last] + ['TPX'] + last
+                           if fs not in last] + ['TPQ'] + last
 
         # determine whether complete concentration is available (mass or mole)
-        # assumes that `X` or `Y` is always in last place
+        # assumes that 'X' or 'Y' is always in last place
+        # (for PureFluid, 'Q_H2O' is treated in a similar fashion)
         mode = ''
-        for prefix in ['X_', 'Y_']:
+        for prefix in ['X_', 'Y_', 'Q_']:
```
Comment (speth) [inline interfaces/cython/cantera/composite.py]: Given that PureFluid is a single-phase species, and I don't think we've worked out an equivalent to the quality for multi-species two-phase mixtures, I think the quality can just be stored as `Q` rather than `Q_speciesName`.

File: interfaces/cython/cantera/test/test_purefluid.py [hunk 3]
```diff
@@ -142,25 +144,88 @@ def test_thermal_expansion_coeff_TD(self):
...
+            self.assertTrue(issubclass(w[-1].category, DeprecationWarning))
+            self.assertTrue("renamed to 'TQ'" in str(w[-1].message))
+
+            X = self.water.X
+            self.assertTrue(issubclass(w[-1].category, DeprecationWarning))
+            self.assertTrue("renamed to 'Q'" in str(w[-1].message))
+
+            self.water.X = X
+            self.assertTrue(issubclass(w[-1].category, DeprecationWarning))
+            self.assertTrue("renamed to 'Q'" in str(w[-1].message))
```
Comment (speth) [inline interfaces/cython/cantera/test/test_purefluid.py]: This test would succeed even if the statement `self.water.X = X` did not issue a `DeprecationWarning`. Ditto for the other paired warnings. That said, I'm not sure it's really necessary to test every single deprecation warning.

File: interfaces/cython/cantera/test/test_purefluid.py [hunk 4]
```diff
@@ -142,25 +144,88 @@ def test_thermal_expansion_coeff_TD(self):
...
+    def test_deprecated_X(self):
+
+        with warnings.catch_warnings(record=True) as w:
+
+            # cause all warnings to always be triggered.
+            warnings.simplefilter("always")
+
+            self.water.TX = 400, 0.8
+            self.assertTrue(issubclass(w[-1].category, DeprecationWarning))
+            self.assertTrue("renamed to 'TQ'" in str(w[-1].message))
```
Comment (speth) [inline interfaces/cython/cantera/test/test_purefluid.py]: Please use the specialized asserts rather than `assertTrue` where possible.

File: interfaces/cython/cantera/thermo.pyx [hunk 5]
```diff
@@ -810,6 +810,19 @@ cdef class ThermoPhase(_SolutionBase):
...
+    property Q:
+        """
+        Placeholder for derived classes with phase transition
+        """
+        def __get__(self):
+            raise NotImplementedError("Attribute 'Q' (vapor fraction) is "
+                                      "not implemented/defined for this phase")
+        def __set__(self, Q):
+            raise NotImplementedError("Attribute 'Q' (vapor fraction) is "
+                                      "not implemented/defined for this phase")
```
Comment (speth) [inline interfaces/cython/cantera/thermo.pyx]: I would like to avoid introducing this kind of placeholder. I think the only reason it's needed is for how the `Q` property is handled in `SolutionArray`, but that could be treated like we do other properties that only appear in derived classes, e.g. `InterfacePhase.coverages`.

**Review by speth — CHANGES_REQUESTED** [review]
I'm fully on board with this renaming. I think it will be worth it to avoid the confusion caused by the overloading of X as both mole fraction and vapor fraction depending on the phase type.

File: interfaces/cython/cantera/composite.py (hunk 2 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: ~With this PR, the mole fraction `X_H2O` will become accessible again (instead of being masked by the quality.~ (actually not, as exceptions are raised). Still, for consistency, I'd prefer to keep the old behavior (it's also much easier to follow the implementation as it just uses what's already there for `X_` and `Y_`). `SolutionArray` will add the species name for other single species mass/mole fractions (which are by definition 1). So if this is done, it applies to `X`, `Y` and `Q` regardless.

File: interfaces/cython/cantera/test/test_purefluid.py (hunk 3 above)
Comment (ischoegl) [reply interfaces/cython/cantera/test/test_purefluid.py]: Yes. Tried to compress this and allowed for that behavior. Fixed.

File: interfaces/cython/cantera/composite.py (hunk 2 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: see:
https://github.com/Cantera/cantera/blob/54c12fc59cfd10a6da2bb7cb81980cd1f7a24696/interfaces/cython/cantera/composite.py#L796-L797

Comment (speth) [reply interfaces/cython/cantera/composite.py]: I think if you treat `Q` like another scalar variable (still unique to the `PureFluid` class, of course), rather than an "n_species" variable, this should work out. I don't have any problem with the per-species variables including `X` and `Y` still ending up like `X_speciesName`.

**Review by speth — COMMENTED** [review]
While trying to experiment with getting `Q` to work as the column name for `SolutionArray` output, I ran into a problem, and haven't quite been able to figure out why it's failing:
```python
>>> import cantera as ct
>>> p = ct.Water()
>>> s = ct.SolutionArray(p, 5)
>>> data, cols = s.collect_data()
>>> s.restore_data(data, cols)
```
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-02284ff2df78> in <module>()
----> 1 s.restore_data(data, cols)

~/src/cantera/build/python/cantera/composite.py in restore_data(self, data, labels)
    704             raise ValueError(
    705                 "invalid/incomplete state information (detected "
--> 706                 "partial information as mode='{}')".format(mode)
    707             )
    708

ValueError: invalid/incomplete state information (detected partial information as mode='Y')
```

**Comment by ischoegl** [comment]
Hm, I can reproduce this. I will have to look at this, as it looks like the standard `TD` state (which is the default) is causing trouble. (Also: likely a pre-existing condition unrelated to the switch to `Q`.)
```python
In [1]: import cantera as ct
   ...: p = ct.Water()
   ...: s = ct.SolutionArray(p, 5)
   ...: data, cols = s.collect_data()

In[2]: cols
Out[2]: ['T', 'density', 'Y_H2O']
```
Will add some additional unit tests to make sure that there aren't any surprises.

**Comment by ischoegl** [comment]
@speth ... on second look, the fix is already included in #720, which allows for different standard states (up until now, the standard state is defined as `TDY` regardless of the phase model; PR #720 sets it to `TD` for `PureFluid`).

So, would you mind tabling this until #720 is discussed & merged? (I recall you mentioning that there may be several issues that may need to be converged).

**Comment by speth** [comment]
Ah, I didn't realize that it was a pre-existing issue, but that makes sense, and I'm happy to see that it's already fixed by #720.

**Comment by ischoegl** [comment]
@speth ... I confirmed that the bug you found was indeed pre-existing, but also found a second issue that prevented #720 from fixing it directly (the additional fix is now added in ce424c9). This is added to #720, as a fix of #644 (and associated issues) is required regardless. I also added a unit test to confirm that things are working as expected.

**Comment by ischoegl** [comment]
Adding an observation from #720 here (per @speth's findings)
```python
gas = ct.Solution('gri30.yaml')
w = ct.Water()
a = ct.SolutionArray(gas, 10)
'TX' in dir(a)  # False, as expected
b = ct.SolutionArray(w)
'TX' in dir(b)  # True, as expected
'TX' in dir(a)  # True, as expected, but not desired
a.TX  # unexpectedly (but as desired) raises AttributeError
```
Once `TX` and `TQ` are differentiated, all should have the expected/desired outcome.

**Comment by ischoegl** [comment]
**Remaining steps:**

 * [x] rebase to incorporate new features introduced in #720 
 * [x] ensure that `data, cols = s.collect_data()` yields `['T', 'density']` 
 * [x] enforce single-letter property acronyms by changing signature of `Phase::nativeState` from `std::map<std::string, size_t>` to `std::map<char, size_t>` 
 * [x] confirm that `SolutionArray` has correct setters (i.e. includes `TPQ` etc.)
 * [x] introduce `Phase::isFluid` flag to facilitate identification of substances with phase change

File: interfaces/cython/cantera/composite.py (hunk 2 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: I believe there is a future use case for multi-component fluids (i.e. *not* pure fluids), that will have dissimilar vapor pressures (and thus different `Q` values). From that perspective, I'd suggest to keep the species name attached and treat vapor fraction as an `_n_species` variable.

**Comment by ischoegl** [comment]
@speth after rebasing to #720, `SolutionArray` appears to be working as expected.
```
In [1]: import cantera as ct
   ...: gas = ct.Solution('gri30.yaml')
   ...: w = ct.Water()
   ...: a = ct.SolutionArray(gas, 10)
   ...: b = ct.SolutionArray(w, 2)
   ...: 

In [2]: [st in dir(w) for st in ['TX', 'TQ']] # expected
Out[2]: [True, True]

In [3]: [st in dir(a) for st in ['TX', 'TQ']] # expected
Out[3]: [False, False]

In [4]: [st in dir(b) for st in ['TX', 'TQ']] # unexpected? <<<<<<<<<<<<<<<<<
Out[4]: [False, False]

In [5]: b.TQ # expected
Out[5]: (array([ 300.,  300.]), array([ 0.,  0.]))

In [6]: b.TQ = 400, .5

In [7]: b.TQ
Out[7]: (array([ 400.,  400.]), array([ 0.5,  0.5]))
```
I.e. I believe this is ready for a review.

**Review by speth — CHANGES_REQUESTED** [review]
Currently, this eliminates the the `TX` and `PX` attributes of `SolutionArray` without a (runtime) deprecation warning. I also noticed that the `X` and `TPX` properties of `SolutionArray` don't seem to emit any warnings for pure fluids.

I'm not really sold on making `nativeState` work with `char` instead of `std::string`. The efficiency benefit shouldn't ever matter, and I don't want to assume that there will always be a clear one-character name for any state variable. For example, what letter would you use for the electron temperature in a non-thermal plasma?

The idea of having `Q` be dimensioned as `nSpecies` to support more general multi-phase fluids is an interesting one, but it's a bit inconsistent here -- the value returned by `PureFluid.Q` is a scalar, while `SolutionArray.Q` returns an array with `nSpecies` in the last dimension, and the second value returned by `SolutionArray.TQ` is not. I realize that this is an inconsistency inherited from the previous mis-use of  `X`, but that's not a great reason to keep it. I'd like to avoid making the interface to the simple case that we already have (i.e. single-component fluids) less convenient in support of a capability that is mostly theoretical at present (would Q even be a useful state setter in such a case?).

Regarding `isFluid`, I'm not really sure I see what the use case is, but if we really want such a flag, then the name is going to need some work. `isMultiphase` would be one option, but I'd worry about distinguishing between "this model includes phase changes" and "the current state involves multiple phases".

**Comment by ischoegl** [comment]
@speth ... thanks for the review. I'm responding to your points before addressing major issue.

> this eliminates the the `TX` and `PX` attributes of `SolutionArray` without a (runtime) deprecation warning. I also noticed that the `X` and `TPX` properties of `SolutionArray` don't seem to emit any warnings for pure fluids.

Yes, I just confirmed this. I hadn't anticipating this as I thought that the unit tests would catch - turns out that I was only testing deprecation of `PureFluid`. Need to pick over 750a80b to see what's going on.

> I'm not really sold on making `nativeState` work with `char` instead of `std::string`. 

I dropped the commit.

> The idea of having `Q` be dimensioned as `nSpecies` to support more general multi-phase fluids is an interesting one, but it's a bit inconsistent here [...] I'd like to avoid making the interface to the simple case that we already have (i.e. single-component fluids) less convenient [...]

I want to believe that Cantera should be able to handle realistic multi-fuel vaporization applications eventually. My thought here was to exactly replicate the behavior of mass/mole fractions. I don't see much of an inconvenience as the only instance where `Q_H2O` etc. is visible is in HDF/CSV output. (It'd still be easy to suppress it for a *pure* substance, while keeping it as `n_species`. Since `X` and `Y` are gone for pure species input/output, there isn't any inconsistency.)

> Regarding `isFluid`, I'm not really sure I see what the use case is, but if we really want such a flag, then the name is going to need some work.

I had `PureFluid` in mind, which would have both `isPure` and `isFluid` set to `true`. I do, however, see your point, and would just name it `hasPhaseTransition`. The reason why I'd like to see this flag being added is that for various equations of state, it would be immediately clear whether phase transitions are implemented or not. Some real gas EOS are capable of predicting the regions, but I wouldn't assume that all cases are covered.

**Comment by speth** [comment]
> I don't see much of an inconvenience as the only instance where `Q_H2O` etc. is visible is in HDF/CSV output.

I don't think `Q_H2O` showing up in the CSV/HDF5 output is particularly problematic. What I'm more concerned about is what is returned by `Solution.Q` when the underlying phase is, say, either `pure-fluid` or `Redlich-Kwong`. Currently, this is a `float` for the former, while the latter is not implemented even at the C++ level, but if were, would presumably need to return an array of length `n_species`, But I don't think we want the return type to vary with the number of species, and I'd rather not have to write `water.Q[0]` to get the vapor fraction.

I don't think the vapor fraction of each species in a multiphase mixture can be used as part of a state setter in the same way as it can for a single-species phase, since it would over-specify the system per the Gibbs' Phase Rule, so I see this as more of a calculated property than part of the state. It could make sense to have two separate properties, one for the total vapor fraction across all species and one for the fraction of each species in the vapor phase. This would be analogous to what we do for many other properties, e.g. `partial_molar_enthalpies` versus `enthalpy_mass`, where the latter is available as part of state setters like `HP`.

> The reason why I'd like to see this flag being added is that for various equations of state, it would be immediately clear whether phase transitions are implemented or not

OK, I can see that being useful for cases like `WaterSSTP` where presumably the correct return value is `false`.

**Comment by ischoegl** [comment]
> I don't think the vapor fraction of each species in a multiphase mixture can be used as part of a state setter in the same way as it can for a single-species phase, since it would over-specify the system per the Gibbs' Phase Rule, so I see this as more of a calculated property than part of the state. 

This is a valid point: a scalar it is.

> ... `WaterSSTP`, `Redlich-Kwong`, etc.

Exactly - it's often not immediately clear what's implemented. I hope we'll be able to set `hasPhaseTransition` to true for more than just `PureFluid` eventually (same for the pending `Peng-Robinson` PR).

**Comment by ischoegl** [comment]
@speth ... per discussion, I switched `isFluid` to `hasPhaseTransition` and made `Q` a scalar.

~The remaining issue is that it appears that setters/getters defined by~
```
getter, setter = _state2_prop(name, Solution)
```
~don't overload in derived classes? Will get back to it next week.~

File: interfaces/cython/cantera/composite.py (hunk 2 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: resolved ... made `Q` a scalar

**Comment by ischoegl** [comment]
@speth ... I believe 174c9b7 addresses setters/getters not being defined correctly, i.e. almost all points previously raised are now addressed.

The lack of warnings is curious: I verified that the functions that should raise the warnings are indeed called. I found another context where warnings (and even errors!) are not raised, which *may* be related: see #746

**Comment by bryanwweber** [comment]
@ischoegl For warnings at least, the warning is only shown the first time a `method` is passed... i.e., the signature of the function is `warn(method, message)`, where the `method` is stored and the message is only shown the first time that `method` is encountered. This is done in C++ so applies to all the interfaces. Not sure if that's the issue you're encountering (and certainly, this is not the resolution to #746).

**Comment by ischoegl** [comment]
>  For warnings at least, the warning is only shown the first time a method is passed...

Correct. The warnings here are, however, raised in `pyx` files, so the issue cannot be caused by the C++ implementation.

**Comment by ischoegl** [comment]
@bryanwweber and @speth ... documenting the lack of warning (which I believe to be a bug that goes beyond this PR)
```
In [1]: import cantera as ct
   ...: gas = ct.Solution('gri30.yaml')
   ...: w = ct.Water()
   ...: w.TQ = 300., .5
   ...: a = ct.SolutionArray(gas, 10)
   ...: b = ct.SolutionArray(w, 2)
   ...: 

In [2]: b.TPX # <--- no warning issued
Out[2]: 
(array([ 300.,  300.]), array([ 3528.21380229,  3528.21380229]), array([[ 0.5],
        [ 0.5]]))

In [3]: w.TPX # <--- calling the method raising the warning directly
/usr/bin/ipython3:1: DeprecationWarning: To be removed after Cantera 2.5. Attribute renamed to 'TPQ'
  #! /bin/sh
Out[3]: (300.0, 3528.213802290444, 0.5000000000000002)

In [4]: b.TPQ
Out[4]: 
(array([ 300.,  300.]),
 array([ 3528.21380229,  3528.21380229]),
 array([ 0.5,  0.5]))
```

File: interfaces/cython/cantera/test/test_composite.py [hunk 6]
```diff
@@ -36,6 +36,11 @@ def test_load_thermo_models(self):
                 self.assertEqual(sol.T, T0)
                 self.assertEqual(sol.P, p0)
 
+                if sol.thermo_model in ('PureFluid'):
```
Comment (speth) [inline interfaces/cython/cantera/test/test_composite.py]: While this happens to work, `('PureFluid')` isn't a tuple -- it should be `('PureFluid',)` instead.

File: interfaces/cython/cantera/test/test_thermo.py [hunk 7]
```diff
@@ -1533,6 +1534,20 @@ def test_get_state(self):
         self.assertEqual(P.shape, (4,))
         self.assertEqual(Y.shape, (4, self.gas.n_species))
 
+    @unittest.expectedFailure
+    def test_idealgas_getters(self):
```
Comment (speth) [inline interfaces/cython/cantera/test/test_thermo.py:1536]: I'm having trouble understanding this test. Why is it an expected failure? What behavior is this test trying to test?

File: interfaces/cython/cantera/composite.py [hunk 8]
```diff
@@ -538,40 +531,20 @@ def __init__(self, phase, shape=(0,), states=None, extra=None):
...
         else:
             raise AttributeError("'{}' object has no attribute '{}'".format(
                 self.__class__.__name__, name))
 
     def __setattr__(self, name, value):
-        if name in self._setters:
-            self._setters[name](self, value)
-        else:
-            super().__setattr__(name, value)
+        super().__setattr__(name, value)
```
Comment (speth) [inline interfaces/cython/cantera/composite.py]: At this point, the `__setattr__` override can just be deleted, right? Likewise, the initialization of `self._extra` can be restored to its simpler form.

File: interfaces/cython/cantera/composite.py [hunk 9]
```diff
@@ -710,16 +680,17 @@ def restore_data(self, data, labels):
                 states = [v[:-1] for v in states if mode in v]
                 break
         if mode == '':
-            # concentration specifier ('X' or 'Y') is not used
-            states = [s for s in states if 'X' not in s and 'Y' not in s]
+            # concentration/quality specifier ('X' or 'Y') is not used
+            states = [st.translate({ord(i): None for i in 'XY'})
+                      for st in states]
```
Comment (speth) [inline interfaces/cython/cantera/composite.py]: A simpler alternative:
```python
states = [st.rstrip('XY') for st in states]
```

File: interfaces/cython/cantera/thermo.pyx [hunk 10]
```diff
@@ -1647,76 +1707,183 @@ cdef class PureFluid(ThermoPhase):
...
+        def __get__(self):
+            return self.u, self.v, self.Q
 
     property DPX:
+        """Get the density [kg/m^3], pressure [Pa], and vapor fraction.
+
+        .. deprecated:: 2.5
+
+             To be deprecated with version 2.5, and removed thereafter.
+             Renamed to `DPQ`.
```
Comment (speth) [inline interfaces/cython/cantera/thermo.pyx]: Would it be worth noting in these deprecation messages that the change in behavior after Cantera 2.5 won't be that this property doesn't exist, but that it will return the mole fraction instead of the vapor fraction? Likewise for the other properties that are shared with `ThermoPhase`.

**Review by speth — CHANGES_REQUESTED** [review]
Thanks for the updates. This is looking pretty good to me.

I think there are some examples that need to be updated to use these new properties. `rankine.py` comes to mind immediately, but there may be others.

File: interfaces/cython/cantera/test/test_thermo.py (hunk 7 above)
Comment (ischoegl) [reply interfaces/cython/cantera/test/test_thermo.py:1536]: This ensures that all getters can actually be used, i.e.
```python
        getters = 'TDPUVHSXY'
        all_getters = [k for k in states.__dir__()
                       if not set(k) - set(getters) and len(k)>1]
```
this would not be true for a setter `TQ`, which is part of `states__dir__()`, but will fail with an attribute error. I added comments.

In a nutshell, these unit tests probe edge cases that previously caused trouble (where getters/setters did not work as expected).

File: interfaces/cython/cantera/composite.py (hunk 8 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: absolutely.

**Comment by ischoegl** [comment]
@speth ... made all requested changes and updated `rankine.py` (couldn't find other examples that are affected)

File: interfaces/cython/cantera/test/test_thermo.py (hunk 7 above)
Comment (speth) [reply interfaces/cython/cantera/test/test_thermo.py:1536]: OK, that makes sense, but I don't think that's what this test achieves. What will happen here is that the first property that works will not raise an `AttributeError` and then test will fail. If you want to test all the getters that should work, you could build that list using the properties of the `ThermoPhase` class. Then you can invert the structure of the test by just eliminating `with` clause.

I think that we should reserve `expectedFailure` for tests that should pass but do not because of a known bug. The only cases where we currently use `expectedFailure` are for some of the equilibrium tests, for which I believe there are open issues in the tracker (although I realize there is unfortunately no comment linking to the relevant issue).

I would also suggest using `dir(foo)` over `foo.__dir__()` and `getattr(foo, bar)` over `foo.__getattribute__(bar)`.

Comment (ischoegl) [reply interfaces/cython/cantera/test/test_thermo.py:1536]: > I think that we should reserve `expectedFailure` for tests that should pass but do not because of a known bug.

Ok. I changed the tests accordingly. 

> What will happen here is that the first property that works will not raise an `AttributeError` and then test will fail. 

Indeed! Thanks for pointing that out - on the upside, the unit test surfaced an old typo (`self.p` instead of `self.P` in `PureFluid.PV`).

**Review by speth — APPROVED** [review]
